insecure/attacknetwork/attackNetwork.go

```diff
@@ -0,0 +1,244 @@
+				if err = s.Serve(ln); err != nil { // blocking call
+					ctx.Throw(fmt.Errorf("could not bind attackNetwork to the tcp listener: %w", err))
+				}
+			}()
+
+			// waits till gRPC server starts serving.
+			wg.Wait()
+			ready()
+		}).
+		AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+			err := attackNetwork.start(ctx)
+			if err != nil {
+				ctx.Throw(fmt.Errorf("could not start attackNetwork: %w", err))
+			}
+
+			ready()
+
+			<-ctx.Done()
+
+			err = attackNetwork.stop()
+			if err != nil {
+				ctx.Throw(fmt.Errorf("could not stop attackNetwork: %w", err))
+			}
+		}).Build()
+
+	attackNetwork.Component = cm
+	attackNetwork.cm = cm
+
+	return attackNetwork, nil
+}
+
+// start triggers the sub-modules of attack network.
+func (a *AttackNetwork) start(ctx irrecoverable.SignalerContext) error {
+	// creates a connection to all corrupted nodes in the attack network.
+	for _, corruptedId := range a.corruptedIds {
+		connection, err := a.corruptedConnector.Connect(ctx, corruptedId.NodeID)
+		if err != nil {
+			return fmt.Errorf("could not establish corruptible connection to node %x: %w", corruptedId.NodeID, err)
+		}
+		a.corruptedConnections[corruptedId.NodeID] = connection
+	}
+
+	// registers attack network for orchestrator.
+	a.orchestrator.WithAttackNetwork(a)
+
+	return nil
+}
+
+// stop conducts the termination logic of the sub-modules of attack network.
+func (a *AttackNetwork) stop() error {
+	// tears down connections to corruptible nodes.
+	var errors *multierror.Error
+	for _, connection := range a.corruptedConnections {
+		err := connection.CloseConnection()
+
+		if err != nil {
+			errors = multierror.Append(errors, err)
+		}
+	}
+
+	// tears down the attack server itself.
+	a.server.Stop()
+
+	return errors.ErrorOrNil()
+}
+
+// ServerAddress returns the address on which the orchestrator is reachable from corrupted nodes.
+func (a AttackNetwork) ServerAddress() net.Addr {
+	return a.address
+}
+
+// Observe implements the gRPC interface of the attack network that is exposed to the corrupted conduits.
+// Instead of dispatching their messages to the networking layer of Flow, the conduits of corrupted nodes
+// dispatch the outgoing messages to the attack network by calling the Observe method of it remotely.
+func (a *AttackNetwork) Observe(stream insecure.Attacker_ObserveServer) error {
+	for {
+		select {
+		case <-a.cm.ShutdownSignal():
+			// attack network terminated, hence terminating this loop.
+			return nil
+		default:
+			msg, err := stream.Recv()
+			if err == io.EOF {
+				a.logger.Info().Msg("attack network closed processing stream")
+				return stream.SendAndClose(&empty.Empty{})
+			}
+			if err != nil {
+				return fmt.Errorf("could not read corrupted node's stream: %w", err)
+			}
+
+			if err = a.processMessageFromCorruptedNode(msg); err != nil {
+				a.logger.Fatal().Err(err).Msg("could not process message of corrupted node")
+				return stream.SendAndClose(&empty.Empty{})
```

Comment:
this will never be called since `Fatal()` will call `os.Exit(1)`. maybe instead of a fatal error, return a custom error type that the caller can use to decide what to do. alternatively, you could accept an `irrecoverable.SignalerContext` and throw the error.

Reply:
Thanks for pointing it out. 

This code is part of the `Observe` method, which implements a client streaming gRPC interface. The interface is generated by compiling a `.proto` file, and cannot be refactored to accept an `irrecoverable.SignalerContext`. Main reason is this `Observe` method is supposed to get called by the gRPC stub solely. 

> this will never be called since `Fatal()` will call `os.Exit(1)`.  

Correct, so the purpose of the code is to safely crash the attacker process in the event that messages arriving from a (remote) corrupted node is not successfully getting through the process path. On a happy path, the attacker must be able to process every incoming message of remote corrupted nodes. If the happy path of processing an incoming message is broken, the testing framework should not continue, and we do it by crashing the attacker. Since attacker is running on the main process that runs the test, crashing it through `os.Exit(1)` will crash the test with the fatal level log. 

The subsequent return was _initially_ there to let the code compile (i.e., all paths of the method ends with a return). However, realized that we no longer need it with current shape of the code, so removed it in https://github.com/onflow/flow-go/pull/2144/commits/3a05692dba6ab4bfe20278bf029773774804efa5